File: src/diagnostics_logger/diagnostics_module_logger.cpp
#include "diagnostics_logger/diagnostics_module_logger.hpp"

namespace auto_battlebot
{
    DiagnosticsModuleLogger::DiagnosticsModuleLogger(const std::string &logger_name) : logger_name_(logger_name),
                                                                                       level_(diagnostic_msgs::DiagnosticStatus::OK),
                                                                                       data_(std::map<std::string, DiagnosticsData>{})
    {
    }

    void DiagnosticsModuleLogger::log(
        int8_t level,
        const std::string &subsection_name,
        const DiagnosticsData &data,
        const std::string &message)
    {
        // Update level to the highest severity
        level_ = std::max(level, level_);

        // Add message if not empty and not already logged
        if (!message.empty() && logged_messages_.find(message) == logged_messages_.end())
        {
            logged_messages_.insert(message);
            messages_.push_back(message);
        }

        // Get or create subsection data map
        if (data_.find(subsection_name) == data_.end())
        {
            data_[subsection_name] = DiagnosticsData{};
        }

        // Combine diagnostic data with existing data in the subsection
        for (const auto &[key, value] : data)
        {
            data_[subsection_name][key] = value;
        }
    }

    void DiagnosticsModuleLogger::debug(const DiagnosticsData &data)
    {
        debug("", data, "");
    }

    void DiagnosticsModuleLogger::debug(const std::string &message)
    {
        debug("", {}, message);
    }

    void DiagnosticsModuleLogger::debug(const std::string &subsection_name, const DiagnosticsData &data)
    {
        debug(subsection_name, {}, "");
    }

    void DiagnosticsModuleLogger::debug(const std::string &subsection_name, const std::string &message)
    {
        debug(subsection_name, {}, message);
    }

    void DiagnosticsModuleLogger::debug(const std::string &subsection_name, const DiagnosticsData &data, const std::string &message)
    {
        log(diagnostic_msgs::DiagnosticStatus::OK, subsection_name, data, message);
    }

    void DiagnosticsModuleLogger::info(const DiagnosticsData &data)
    {
        info("", data, "");
    }

    void DiagnosticsModuleLogger::info(const std::string &message)
    {
        info("", {}, message);
    }

    void DiagnosticsModuleLogger::info(const std::string &subsection_name, const DiagnosticsData &data)
    {
        info(subsection_name, {}, "");
    }

    void DiagnosticsModuleLogger::info(const std::string &subsection_name, const std::string &message)
    {
        error(subsection_name, {}, message);
    }

    void DiagnosticsModuleLogger::info(const std::string &subsection_name, const DiagnosticsData &data, const std::string &message)
    {
        log(diagnostic_msgs::DiagnosticStatus::OK, subsection_name, data, message);
    }

    void DiagnosticsModuleLogger::warning(const DiagnosticsData &data)
    {
        warning("", data, "");
    }

    void DiagnosticsModuleLogger::warning(const std::string &message)
    {
        warning("", {}, message);
    }

    void DiagnosticsModuleLogger::warning(const std::string &subsection_name, const DiagnosticsData &data)
    {
        warning(subsection_name, {}, "");
    }
    void DiagnosticsModuleLogger::warning(const std::string &subsection_name, const std::string &message)
    {
        error(subsection_name, {}, message);
    }
    void DiagnosticsModuleLogger::warning(const std::string &subsection_name, const DiagnosticsData &data, const std::string &message)
    {
        log(diagnostic_msgs::DiagnosticStatus::WARN, subsection_name, data, message);
    }

    void DiagnosticsModuleLogger::error(const DiagnosticsData &data)
    {
        error("", data, "");
    }

    void DiagnosticsModuleLogger::error(const std::string &message)
    {
        error("", {}, message);
    }

    void DiagnosticsModuleLogger::error(const std::string &subsection_name, const DiagnosticsData &data)
    {
        error(subsection_name, {}, "");
    }
    void DiagnosticsModuleLogger::error(const std::string &subsection_name, const std::string &message)
    {
        error(subsection_name, {}, message);
    }
    void DiagnosticsModuleLogger::error(const std::string &subsection_name, const DiagnosticsData &data, const std::string &message)
    {
        log(diagnostic_msgs::DiagnosticStatus::ERROR, subsection_name, data, message);
    }

    void DiagnosticsModuleLogger::clear()
    {
        level_ = diagnostic_msgs::DiagnosticStatus::OK;
        messages_.clear();
        data_.clear();
        logged_messages_.clear();
    }

    bool DiagnosticsModuleLogger::has_status() const
    {
        return !messages_.empty() || !data_.empty();
    }

    std::vector<diagnostic_msgs::DiagnosticStatus> DiagnosticsModuleLogger::get_status() const
    {
        // Join messages with " | " separator
        std::string combined_message;
        for (size_t i = 0; i < messages_.size(); ++i)
        {
            if (i > 0)
            {
                combined_message += " | ";
            }
            combined_message += messages_[i];
        }

        std::vector<diagnostic_msgs::DiagnosticStatus> diagnostics;
        for (const auto &[subsection_name, data_map] : data_)
        {

            diagnostics.push_back(dict_to_diagnostics(
                data_map,
                level_,
                subsection_name,
                combined_message,
                logger_name_));
        }
        return diagnostics;
    }

    std::string DiagnosticsModuleLogger::get_name() const
    {
        return logger_name_;
    }

} // namespace auto_battlebot

File: playground/diagnostics_module_logger.py
from diagnostic_msgs.msg import DiagnosticStatus

from app.shared.messages.diagnostics.dict_to_diagnostics import dict_to_diagnostics


class DiagnosticsModuleLogger:
    """
    A logger for a specific module that accumulates diagnostic information. This should
    not be used directly; instead, use DiagnosticsLogger to manage multiple module loggers.
    """

    def __init__(self, app_name: str, logger_name: str) -> None:
        self.app_name = app_name
        self.logger_name = logger_name
        self.data: dict = {}
        self.level = DiagnosticStatus.OK
        self.messages: list[str] = []
        self._logged_messages: set[str] = set()
        self.clear()

    def log(self, level: int, data: dict | None = None, message: str = "") -> None:
        self.level = max(level, self.level)
        if message and message not in self._logged_messages:
            self._logged_messages.add(message)
            self.messages.append(message)
        if data:
            self.data.update(data)

    def debug(self, data: dict | None = None, message: str = "") -> None:
        self.log(DiagnosticStatus.OK, data, message)

    def info(self, data: dict | None = None, message: str = "") -> None:
        self.log(DiagnosticStatus.OK, data, message)

    def warning(self, data: dict | None = None, message: str = "") -> None:
        self.log(DiagnosticStatus.WARN, data, message)

    def error(self, data: dict | None = None, message: str = "") -> None:
        self.log(DiagnosticStatus.ERROR, data, message)

    def clear(self) -> None:
        self.level = DiagnosticStatus.OK
        self.messages = []
        self.data = {}
        self._logged_messages = set()

    def has_status(self) -> bool:
        return len(self.messages) > 0 or len(self.data) > 0

    def get_status(self) -> DiagnosticStatus:
        message = " | ".join(self.messages) if self.messages else ""
        return dict_to_diagnostics(
            self.data, self.level, self.logger_name, message, self.app_name
        )
